projects/transformers/callbacks/sparsity.py

```diff
@@ -75,3 +93,266 @@ def on_step_end(self, args, state, control, model, **kwargs):
             )
 
             wandb.log(logs, step=state.global_step)
+
+
+class RigLCallback(TrainerCallback):
+    """
+    Perform `RigL`_ dynamic sparsity every N steps on the sparse modules within the
+    network. Weights are pruned according to their ranks absolute values and regrowing
+    (adding weights) occurs by ranking their absolute gradients. As per the original
+    paper, the pruning decreases over time according to a cosine decay.
+
+    .. _RigL: https://arxiv.org/abs/1911.11134
+
+    :param prune_fraction: initial fraction of weights to prune
+    :param prune_freq: how often, in iterations, to prune and regrow weights
+    :param warmup_steps: defaults to prune_freq; e.g. prune_freq of 100 will allow 100
+                         steps before pruning for the first time
+    """
+
+    def __init__(
+        self,
+        prune_fraction=0.3,
+        prune_freq=100,
+        warmup_steps=None
+    ):
+        self.prune_fraction = prune_fraction
+        self.prune_freq = prune_freq
+        self.prune_scheduler = None
+        self.warmup_steps = warmup_steps
+
+    def on_init_end(self, args, state, control, model, optimizer=None, **kwargs):
+        """Save a list of the sparse modules and initialize the pruning schedule"""
+
+        warmup_steps = self.warmup_steps or self.prune_freq
+        self.prune_scheduler = CosineDecayPruneScheduler(
+            total_steps=args.max_steps,
+            prune_fraction=self.prune_fraction,
+            warmup_steps=warmup_steps - 1
+        )
+        self.sparse_modules = filter_modules(
+            model, include_modules=[SparseWeightsBase]
+        ).values()
+
+    def on_step_end(
+        self, args, state, control, model=None,
+        train_dataloader=None, optimizer=None, **kwargs
+    ):
+        """Prune and regrow weights every 'prune_freq' iterations."""
+
+        if state.global_step % self.prune_freq != 0:
+            self.prune_scheduler.step()
+            return
+
+        # Pre-prune sparsities.
+        param_sparsity0, mask_sparsity0 = calc_cumulative_sparsity(self.sparse_modules)
+
+        # Prune weights.
+        model.apply(rezero_weights)
+        prune_fraction = self.prune_scheduler.get_prune_fraction()
+        num_removed = global_prune_by_weight(self.sparse_modules, prune_fraction)
+        model.apply(rezero_weights)
+
+        # Post-prune sparsities.
+        param_sparsity1, mask_sparsity1 = calc_cumulative_sparsity(self.sparse_modules)
+
+        # Train for one step.
+        optimizer.zero_grad()
+        train_batch = next(iter(train_dataloader))
+        inputs_to_device(train_batch, device=args.device)
+        output = model(**train_batch)
```

Comment:
If compute_loss is not being used, this can't be combined with our distillation mixin.